content/utilities/highlight-friends/plugin.js

```diff
@@ -0,0 +1,278 @@
+	style=${selectStyle}
+	value=${inMinLevel}
+	onChange=${(e) => {
+      minLevel = Number(e.target.value);
+      setInMinLevel(Number(e.target.value))
+    }
+    }
+  >
+	${PLANET_LEVELS.map(
+      ({ value, text }) => html`<option value=${value}>${text}</option>`
+    )}
+  </select>
+`;
+
+  const maxLevelSelect = html`
+<select
+  style=${selectStyle}
+  value=${inMaxLevel}
+  onChange=${(e) => {
+      maxLevel = Number(e.target.value);
+      setInMaxLevel(Number(e.target.value))
+
+    }
+    }
+>
+  ${PLANET_LEVELS.map(
+      ({ value, text }) => html`<option value=${value}>${text}</option>`
+    )}
+</select>
+`;
+
+  let friendChildren =
+    inPlanet.map((item) => {
+      return html`<div>
+		${shortAddress(item.owner)}
+		${' '}
+		<button style=${buttonStyle}
+		 onClick=${() => { center(item) }}
+		> center </button>
+		${' '}
+		<button
+		style=${buttonStyle}
+		onClick=${() => { removeFriend(item); }}
+		> delete </button>
+	  </div>`;
+    });
+
+  function aimPlanet() {
+    let tmpPlanet = ui.getSelectedPlanet();
+    setChosenPlanet(tmpPlanet);
+  }
+
+  function center(thePlanet) {
+    ui.centerLocationId(thePlanet.locationId);
+  }
+
+  function addFriend() {
+    if (chosenPlanet == undefined) return;
+    let flag = true;
+    for (let i in HIGHLIGHT_FRIENDS.planet) {
+      if (HIGHLIGHT_FRIENDS.planet[i].owner == chosenPlanet.owner)
+        flag = false;
+    }
+
+    if (flag == true) HIGHLIGHT_FRIENDS.planet.push(chosenPlanet);
+    setInPlanet([...HIGHLIGHT_FRIENDS.planet]);
+    friend = [];
+    for (const i in HIGHLIGHT_FRIENDS.planet)
+      friend.push(HIGHLIGHT_FRIENDS.planet[i].owner);
+  }
+
+
+  function removeFriend(thePlanet) {
+    let newPlanet = [];
+    for (let i in HIGHLIGHT_FRIENDS.planet) {
+      let p = HIGHLIGHT_FRIENDS.planet[i];
+      if (p.owner == thePlanet.owner) continue;
+      newPlanet.push(p);
+    }
+    HIGHLIGHT_FRIENDS.planet = newPlanet;
+    setInPlanet([...HIGHLIGHT_FRIENDS.planet]);
+    friend = [];
+    for (const i in HIGHLIGHT_FRIENDS.planet) {
+      friend.push(HIGHLIGHT_FRIENDS.planet[i].owner);
+    }
+  }
+
+  function clearFriend() {
+    HIGHLIGHT_FRIENDS.planet = [];
+    friend = [];
+    setInPlanet(HIGHLIGHT_FRIENDS.planet);
+  }
+
+  return html`<div  style=${divStyle} >
+
+	  <button onClick=${aimPlanet}> choose planet </button>
+	  <h1> ${chosenPlanet == undefined ? "not choose yet" : chosenPlanet.owner}</h1>
+
+	  <div> 
+
+	  <button onClick=${addFriend}>add friend</button>
+	  ${' '}
+	  <button onClick=${clearFriend}>clear all</button>
+	  </div>
+	  <div>${' '}</div>
+
+	  <div 
+	  style=${planetLevelStyle}
+
+		> 
+		<h1> choose the planet Level </h1>
+
+		${minLevelSelect}
+		${' '}
+		${maxLevelSelect}
+
+	  </div>
+
+	  <div> number of friends： ${inPlanet.length}</div>
+	  <div style=${listStyle}
+	  > ${friendChildren}</div>
+  </div>`;
+}
+
+
+function App() {
+  return html`<${friendList} />`;
+}
+
+class Plugin {
+  constructor() {
```

Comment:
I get 
```
Failed to start plugin: highlight friends - Please review stack trace
 TypeError: Cannot read properties of undefined (reading 'planet')
    at new Plugin (a2f1eac4-9158-4928-b8bb-74e2269169ef:218)
    at gu.spawn (PluginManager.tsx:208)
    at async gu.render (PluginManager.tsx:223)
spawn @ PluginManager.tsx:211
```

I think HIGHLIGHT_FRIENDS doesnt exist so you cant create HIGHLIGHT_FRIENDS.planet?

Reply:
I will make a little change, and commit a new version soon.
actually change 'window.HIGHLIGHT_FRIENDS.planet' to 'window.HIGHLIGHT_FRIENDS', because just use one variable .

Reply:
I think you might want more variables in the future without having to break it.

I think you maybe just needed
```
window.HIGHLIGHT_FRIENDS = {}
window.HIGHLIGHT_FRIENDS.planet =[];
```

Reply:
Thanks for your advice! I make change like this, and test this plugin on chrome and firefox (both clean the cache) and it runs well.